src/components/input/SelectNext.tsx

```diff
@@ -0,0 +1,182 @@
+import classnames from 'classnames';
+import type { ComponentChildren } from 'preact';
+import {
+  useCallback,
+  useContext,
+  useId,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from 'preact/hooks';
+
+import { useArrowKeyNavigation } from '../../hooks/use-arrow-key-navigation';
+import { useClickAway } from '../../hooks/use-click-away';
+import { useKeyPress } from '../../hooks/use-key-press';
+import { MenuCollapseIcon, MenuExpandIcon } from '../icons';
+import Button from './Button';
+import SelectContext from './SelectContext';
+
+export type SelectProps<T> = {
+  selected: T;
+  onChange: (newValue: T) => void;
+  label: ComponentChildren;
+  children: ComponentChildren;
+  disabled?: boolean;
+};
+
+function SelectMain<T>({
+  label,
+  selected,
+  onChange,
+  children,
+  disabled,
+}: SelectProps<T>) {
+  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
+  const closeDropdown = useCallback(() => setIsDropdownOpen(false), []);
+  const wrapperRef = useRef<HTMLDivElement | null>(null);
+  const listboxId = useId();
+  const buttonRef = useRef<HTMLButtonElement | null>(null);
+  const buttonId = useId();
+  const [shouldDropUp, setShouldDropUp] = useState(false);
+
+  const selectValue = useCallback(
+    (newValue: unknown) => {
+      onChange(newValue as T);
+      closeDropdown();
+    },
+    [closeDropdown, onChange],
+  );
+
+  // When clicking away or pressing `Esc`, close the dropdown
+  useClickAway(wrapperRef, closeDropdown);
+  useKeyPress(['Escape'], closeDropdown);
+
+  // Vertical arrow key for options in the dropdown
+  useArrowKeyNavigation(wrapperRef, { horizontal: false });
+
+  // Focus button after closing dropdown
+  useLayoutEffect(() => {
+    if (!isDropdownOpen) {
+      buttonRef.current!.focus();
+    }
+  }, [isDropdownOpen]);
+
+  if (isDropdownOpen) {
+    const viewportHeight = window.innerHeight;
+    const { top: buttonDistanceToTop, height } =
+      buttonRef.current!.getBoundingClientRect();
+    const buttonDistanceToBottom =
+      viewportHeight - (buttonDistanceToTop + height);
+
+    // The listbox should drop up only if there's not enough space below it for
+    // the listbox max height, and there's also more space above
+    setShouldDropUp(
+      // 320px is tailwind's h-80. e should try not to couple with this
+      buttonDistanceToBottom < 320 &&
```

Comment:
Where does the 320px (or `h-80` equivalent) come from? In my testing I found that the dropdown was dropping up even in cases where it looked like there was enough room that it should drop down.

Reply:
Yeah, there's room for improvement here.

Since the dropdown is not even rendered, it is not possible to infer how much size it is going to need, so what I did is use the maximum space it could take as reference (notice the listbox has `max-h-80`).

I tried to see how other libraries do this, but since they support other use cases, their code is more complex, and it's not easy to extract the piece solving this specific problem.

But yeah, the main problem with this is that it calculates based on the maximum size, which means, if the listbox is actually smaller, it can end up dropping up even if it would have been able to fit down.

I'll put a bit more thought on this logic. Perhaps moving the logic to `useLayoutEffect` as you pointed out in other comments, ends up solving this as well, but I want to avoid a listbox that "blinks" down and is immediately repositioned after. I'm trying to make it directly appear either up or down.

Reply:
I can confirm that using `useLayoutEffect` + toggling via `display: none` instead of via unmount, allows for the actual height of the listbox to be calculated, making this more accurate.

Also, hidding via CSS should also help addressing https://github.com/hypothesis/frontend-shared/pull/1237#discussion_r1331205890, but this needs to be confirmed.